File: linkedList/lruCache.py
#I watched the neetcode video on how to implement the LRU cache on papar but did the coding implementation myself w/o peaking.
#first time doing OOO coding from scratch and am proud of myself as this worked on the firt try
# This LRU cache implementation uses a hash map and double LL for O(1) get and put functions

class Node(object):
    def __init__(self, key, value):
        self.key = key
        self.value = value
        self.prev = None
        self.next = None
    
class LRUCache(object):
    def __init__(self, capacity):
        """
        :type capacity: int
        """
        #init maxCap, curCap, hashTable and double LL left & right dummy nodes
        self.maxCap = capacity
        self.curCap = 0
        self.hash = {}
        self.left = Node(0,0)
        self.right = Node(0,0)
        self.left.next = self.right
        self.right.prev = self.left

    def removeNode(self, node):
        #create holder node vars for node's prev and next nodes
        prevNode = node.prev
        nextNode = node.next
        #update nodes pointers to point to None
        node.prev = None
        node.next = None
        #update prevNode and nextNode pointers
        prevNode.next = nextNode
        nextNode.prev = prevNode
        #return node for either freeing or inserting
        return node

    def insertNode(self, node):
        #insert Node @ righ most end of LL 
        #create holder vars for right node of LRU cache and right's prev
        rightNode = self.right
        behindRightNode = rightNode.prev
        #update rightNode and behindRightNode to point to node
        rightNode.prev = node
        behindRightNode.next = node
        #update node's prev and next pointers
        node.next = rightNode
        node.prev = behindRightNode

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        #check to see if key is in hash
        if key in self.hash:
            #grab the value
            val = self.hash[key].value
            #move the Nodes to right of the list
            node = self.removeNode(self.hash[key])
            self.insertNode(node)
            #return val
            return val
        #if not, return -1
        return -1

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        # check to see if key exists
        if key in self.hash:
            #update its value
            self.hash[key].value = value
            #move node to right of the LL
            node = self.removeNode(self.hash[key])
            self.insertNode(node)
            return
        #if key doesn't exist, add it
        if self.curCap >= self.maxCap:
            #if the capacity is met, we evict LRU node,
            lruNode = self.removeNode(self.left.next)
            #update hash table
            del self.hash[lruNode.key]
            #add new node
            newNode = Node(key, value)
            self.hash[key] = newNode
            self.insertNode(newNode)
        else:
            #if max capacity isn't met, just add it to cache
            node = Node(key, value)
            self.hash[key] = node
            self.insertNode(node)
            self.curCap += 1
# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: linkedList/lruCache.cpp
class Node {
//
public:
    //
    Node* p;
    Node* n;
    int k;
    int v;
    Node(int key, int value, Node* prev, Node* next){
        p = nullptr;
        n = nullptr;
        k = key;
        v = value;
    }
};

class LRUCache {
public:
    //we need a head & tail pointers 
    Node* h;
    Node* t;
    //LRU capacity var
    int cap;
    //LRU cur size
    int cur_size;

    //hashtable to grab Node addresses based on key
    unordered_map< int, Node* > hash;

    LRUCache(int capacity) {
        // assign capacity and init cur_size
        cap = capacity;
        cur_size = 0;

        //create an empty node to assign head and tail to
        //h = new Node(-1,-1,nullptr,nullptr);
        h = nullptr;
        t = h;
    }
    
    void moveNodeToHead(Node* cur){
        // take key, value Node and put it at the head of LL
        // if cur is head, nothing to do
        if (cur == h){
            return;
        }

        //if this is the first Node in LL, just ass
        if(h == nullptr && t == nullptr){
            h = cur;
            t = cur;
            return;
        }

        // if cur isn't head
        //take care of neighbors pointers
        if (cur -> p){
            cur -> p -> n = cur -> n;
        }
        if (cur -> n){
            cur -> n -> p = cur -> p;
        }

        //take care of cur's points while updating t accordingly
        cur -> n = h;
        if (cur == t){
            t = t -> p;
        }
        cur -> p = nullptr;

        //update h
        if(h){
            h -> p = cur;
        }
        h = cur;
    }

    int get(int key) {
        // if it doesn't exist, return -1
        if(!hash[key])
            return -1;
        
        // if it does exist, we want to move LL val to front of the LL
        moveNodeToHead(hash[key]);
        // then return value
        return hash[key] -> v;
    }
    
    void put(int key, int value) {
        // if it doesn't exist, create new node & update hashtable
        Node* cur;
        if (!hash[key]){
            cur = new Node(key, value, nullptr, nullptr);
            cur_size++;
            hash[key] = cur;
        }
        else{
            cur = hash[key];
            cur -> v = value;
        }
        
        //put cur at front of list
        moveNodeToHead(cur);


        // if size is greater than cap, evict tail:)
        if(cur_size <= cap){
            return;
        }

        //update size
        cur_size--;

        //update hash table
        hash[t -> k] = nullptr;

        //update tail to prev value and delete old tail
        Node* tmp = t;
        t = t -> p;
        t -> n = nullptr;

        tmp -> p = nullptr;
        //free(tmp);
        delete tmp;
    }
};
